File: Samples/TKObjc/TKObjc/ViewController.m
//
//  ViewController.m
//  TKObjc
//
//  Created by Greg Cockroft on 9/27/16.
//  Copyright © 2016 Token Inc. All rights reserved.
//

/*
   Not much yet. Few calls copied from the test project
   Just enough to prove the non-framework Pod works
 */
#import "ViewController.h"
#import "TokenSdk/TokenSdk.h"

@class TokenIO;
@class TKMember;
@class TKAccount;

@interface ViewController ()

@end


@implementation ViewController {
    TKMember *member;
    TokenIO *tokenIO;
    dispatch_queue_t queue;
}


- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self createMember];
}

- (void)setUp {

}

typedef void (^AsyncTestBlock)(TokenIO *);

- (void)run:(AsyncTestBlock)block {
    __block NSException *error;
    
    dispatch_semaphore_t done = dispatch_semaphore_create(0);
    dispatch_async(queue, ^{
        @try {
            block(tokenIO);
        } @catch(NSException *e) {
            error = e;
        } @finally {
            dispatch_semaphore_signal(done);
        }
    });
    
    // Wait for the block above to finish executing while running the main
    // event loop. gRPC posts to the main dispatch queue, so we have to run it
    // for the callbacks to be completed.
    while (true) {
        @try {
            [[NSRunLoop mainRunLoop] runMode:NSDefaultRunLoopMode
                                  beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
        } @catch(NSException *e) {
            NSLog(@"**ERROR: %@", e);
            break;
        } @catch(NSObject *o) {
            NSLog(@"**UNKNOWN ERROR: %@", o);
            break;
        }
        
        // Are we done yet?
        if (dispatch_semaphore_wait(done, DISPATCH_TIME_NOW) == 0) {
            break;
        }
    }
    
    if (error) {
        @throw error;
    }
}


- (TKMember *)createMember:(TokenIO *)token {
    NSString *username = [@"username-" stringByAppendingString:[TKUtil nonce]];
    return [token createMember:username];
}

- (void)createMember
{
    
    TokenIOBuilder *builder = [TokenIO builder];
    builder.host = @"dev.api.token.io";
    builder.port = 90;
    tokenIO = [builder build];
    //bank = [TKBankClient bankClientWithHost:fank.host
      //                                 port:fank.port];
    
    queue = dispatch_queue_create("io.token.Test", nil);
    [self run: ^(TokenIO *tkIO) {
        member = [self createMember:tkIO];
    }];
    
    
}


@end

File: Samples/TKSwift/TKSwift/ViewController.swift
//
//  ViewController.swift
//  TKSwift
//
//  Created by Greg Cockroft on 9/27/16.
//  Copyright © 2016 Token Inc. All rights reserved.
//

import UIKit
import TokenSdk

class ViewController: UIViewController {
    var tokenIOAsync:TokenIOAsync?
    var tkmember:TKMemberAsync?
    var client:TKClient?
    var memberId:String?
    var key:TKSecretKey?
    let gatewayhost = "dev.api.token.io"
    let gatewayport:Int32 = 90
    //
    @IBOutlet weak var memberLabel:UILabel!
    @IBOutlet weak var usernameTextfield: UITextField!

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
    }
    
    
    func info(title:String, message:String) {
        let myAlert = UIAlertController(title: "Info", message: message, preferredStyle: UIAlertControllerStyle.alert)
        myAlert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default) { _ in
        })
        self.present(myAlert, animated: true, completion: nil)
    }
    
    @IBAction func createMember(_ sender : AnyObject) {
        let builder = TokenIOBuilder()!
        print("builder is this \(builder)")
        
        builder.host = gatewayhost
        builder.port = gatewayport
        
        tokenIOAsync = builder.buildAsync()
        print("tokenIOAsync is this \(tokenIOAsync)")
        
        tokenIOAsync?.createMember(nil,
                  onSucess: { (member) -> Void in
                    self.tkmember = member
                    self.memberId = member?.id
                    self.memberLabel.text = "Id: \(self.memberId)"
                    print("member is this \(self.tkmember)")
                    print("member.firstUsername should be nil is this \(self.tkmember?.firstUsername)")
            },
            onError: { (error) -> Void in
                self.info(title:"Error",message:"\(error)")
                print("createMember Error:\(error)")
            }
        )
    
    }
    
    @IBAction func addUsername(_ sender : AnyObject) {
        tkmember?.addUsername(usernameTextfield.text,
                          onSucess: { (member) -> Void in
                            self.info(title:"Info",message:"Done")
           },
                          onError: { (error) -> Void in
                            self.info(title:"Error",message:"\(error)")
                            print("addUsername Error:\(error)")
        })
    }
    
    @IBAction func removeUsername(_ sender : AnyObject) {
        tkmember?.removeUsername(usernameTextfield.text,
                           onSucess: { () -> Void in
                            self.info(title:"Info",message:"Done")
            },
                           onError: { (error) -> Void in
                            self.info(title:"Error",message:"\(error)")
                            print("removeUsername Error:\(error)")
        })
    }
    
    
}

